### ISSUE TITLE
Add a way to transform between block variations

### ISSUE THREAD

**Issue by draganescu** [WordPress/gutenberg#25231]
The navigation block has two variations: horizontal and vertical. These have to be variations because styles are CSS only and we need to alter one of the block's attributes to change how the movers behave depending on orientation.

However this is not very intuitive and it's not easy to understand the difference and the "why", and for many instances its has been suggested to "merge the two variations" (e.g. #23226). As there is no way to "merge" the variations, the proxy would be something which behaves like styles and appears to users as a "setting" of the block.

Therefore it could be so that when a block has multiple variations registered it displays the list of variations in one of the two places occupied currently by styles:

<img width="1134" alt="Screenshot 2020-09-11 at 10 38 43" src="https://user-images.githubusercontent.com/107534/92885025-10b7d980-f41b-11ea-9095-69136985d3f1.png">

The new API should be able to partially apply new block attributes, preserving content.

In the case of the navigation example, starting from:
```
<!-- wp:navigation {"orientation":"vertical"} -->
<!-- wp:navigation-link {"label":"Ut atque aut nobis","type":"custom","id":1976} -->
<!-- wp:navigation-link {"label":"Sample Page","type":"custom","id":1979} /-->
<!-- /wp:navigation -->
```
... after applying the operation the result is:

```
<!-- wp:navigation {"orientation":"horizontal"} -->
<!-- wp:navigation-link {"label":"Ut atque aut nobis","type":"custom","id":1976} -->
<!-- wp:navigation-link {"label":"Sample Page","type":"custom","id":1979} /-->
<!-- /wp:navigation -->
``` 

This new feature is also useful for block based themes as it allows themes to apply their own attribute customizations to existing data saved in template parts.

**Comment by shaunandrews** [comment]
I've explored this briefly, as its something that comes up again and again. The first option that came to mind was to list variations just like styles in the block menu:

<img width="321" alt="image" src="https://user-images.githubusercontent.com/191598/92940372-0c8cbb80-f41d-11ea-8d2c-81f5ae5c9109.png">

This can get rather large, and there's potential for confusion between styles and variations. My second idea was to move styles and variations into a submenu to better differentiate and organize:

<img width="557" alt="image" src="https://user-images.githubusercontent.com/191598/92940481-2d551100-f41d-11ea-8bb2-a06cbc1c33ac.png">

I like this, but we obviously don't have this capability yet with the current menu system. Which led me towards exploring how we could use the existing Transform menu:

<img width="322" alt="image" src="https://user-images.githubusercontent.com/191598/92940603-4cec3980-f41d-11ea-919b-de586836574e.png">

This is fine, and would get the job done, but it lacks the visual emphasis and doesn't really explain the differences between variations.

Moving away from the toolbar, perhaps switching variations isn't a "primary" function and could make more sense in the sidebar:

<img width="651" alt="image" src="https://user-images.githubusercontent.com/191598/92940791-8a50c700-f41d-11ea-8f24-4b56c8e72735.png">

This would work ok, but I wonder if it could be confused with (or should actually be) the default setting for the block across the site.

In any event, I think more exploration needs to go into the above options — with the sub-menu concept being my favorite as the pattern could be useful in many other places in the editor.

**Comment by ntsekouras** [comment]
> Moving away from the toolbar, perhaps switching variations isn't a "primary" function and could make more sense in the sidebar:

I think so too. In general we have to consider that a block could have many variations and the perfect example is the `embed` block with over 20 variations :)

Another thing to consider is that while many variations change attributes only and could be easily understood from their name, others rely on visuals/icons like the `Columns` variations (https://github.com/WordPress/gutenberg/issues/20584).

**Comment by mapk** [comment]
Aren't all the Embeds just variations of the Embed block? If we call these variations as well, does that mean each of these would appear in the Transform Tool? 

![Screen Shot 2020-10-06 at 9 25 33 AM](https://user-images.githubusercontent.com/617986/95229907-0092e000-07b6-11eb-8c01-44a7df1f099f.png)


For this reason, I've chosen to call the sort of changes in this particular issue "Patterns" since they aren't changing the settings of how the block works, but rather just changing the blocks being used. I've been thinking this through in https://github.com/WordPress/gutenberg/issues/25750#issuecomment-704353459. 

A **block variation**, to me, affects the settings of the block. A **pattern** affects the layout of blocks being used within it, not necessarily the settings. Ugh. If we can find a clear differentiator here, I think that can really help provide a clear separation of functionality.

**Comment by ntsekouras** [comment]
> A block variation, to me, affects the settings of the block.

That's true, but there might be display settings as well, that will lead in `layout` changes (check Columns variations). You shouldn't map them mentally to change only attributes that affect the functionality of a block.

A pattern with our current API is more like a template maybe. It's a list of blocks to be inserted. It can contain many different blocks and is not limited to a single block. A pattern could just insert a heading and a paragraph and their purpose is to add content, in the form of a number of blocks, at once.

Block variations have the option to be shown in the inserter or not. For example `embed` shows them and `columns` are not. What is missing here with block variations is the ability to transform a block to a variation of itself, after creation, because now he show only transformation to a different block.

Hope that helps clarify a bit more the difference @mapk.

**Comment by mapk** [comment]
Thanks, @ntsekouras! I think between the terminology and having some variations show in the Inserter while others show in the Transform Tool of the block is what's confusing. Can this be defined further? Like maybe if the block variation is a different block, that should be surfaced in the Inserter (ie. Embeds). But if the block variation is a different version/layout of itself then it is shown in the Transform Tool of the block. 

But this puts the Query block in an interesting position. A block variation might be the Latest Posts block which would show in the Inserter as a different block. But another block variation of the Query block might be shown in the Transform Tool because it's really just changing the layout like grid or list views. Do we call these both variations?

**Comment by draganescu** [comment]
What we're trying to achieve here is to have the mechanism to apply attributes updates to the same content, thus effectively enabling transforming between variations. We should keep this in focus.

I think this discussion is making this issue split into a high level dilemma about when is a block made of blocks a pattern and when is a pattern a block.

If the layout of the query block is decided by setting one of its attributes (say named layout) to some value (say one of top_images, no_images, list, columns) then this is also a transform between variations just like a transform between vertical and horizontal navigation is.

I don't necessarily agree that having the variations showing up in the inverter and in the transform tool is confusing since we're already doing this with block transforms. I do agree that some blocks would have so many variations that the UX would be impossible. What we could do is disable transform variations for some blocks?

**Comment by mapk** [comment]
>I don't necessarily agree that having the variations showing up in the inverter and in the transform tool is confusing since we're already doing this with block transforms.

I would agree. The problem, for me, is around why _some_ variations would show in the Inserter and others in the Transform tool in block's toolbar. Having a clear definition around that would be super helpful.

**Comment by mtias** [comment]
Good explanation @ntsekouras :)

**Comment by gziolo** [comment]
I wanted just second what @ntsekouras said that we can expand the types of supported `scope` values. To give an example, if we would have another type called `transforms`, it would be way much easier to adjust them to fit in the UI where they are presented and how they behave.

In addition to that, we need to ensure users never lost their content, so we can't override inner blocks the way we do it for Columns (and soon Query) block when picking a pattern means injecting a predefined template. Instead, we would have to define a set of attribute updates applied to the block and its children.

**Comment by mapk** [comment]
Looks like we need a decision on a path forward here? I'd really like to [use this interaction on the Query block](https://github.com/WordPress/gutenberg/issues/25750#issuecomment-703851793) too. 

The initial design from @shaunandrews is the one I like best, but it doesn't scale well. It's the same solution I arrived at for the Query block. However I foresee many different layouts (ie. 10+) and if that's the case, should we go with Shaun's second option above?

**Comment by ntsekouras** [comment]
> Looks like we need a decision on a path forward here?

I second that!

>Moving away from the toolbar, perhaps switching variations isn't a "primary" function and could make more sense in the sidebar

I favor the third option for the reason @shaunandrews states. Although I don't think the Block Card would be the perfect place to live. I think it could live in a Panel of its own, having a title/description and probably a fixed width(with scrolling) for cases with many variation transformations.

Lastly a thing to have in mind for the design process is that, as we have discussed the implementation with @gziolo, transforming between block variations will be an opt-in behavior for variations, as it doesn't make sense to show all block variations in every block and additionally there will be cases that the transformation will need a custom `transform` function(especially with `InnerBlocks`).

**Comment by paaljoachim** [comment]
Some variations need a visual thumbnail panel to show a structural change.
An example columns with different layouts. 

Initial variation layout. 
![Screen Shot 2020-10-31 at 11 08 45](https://user-images.githubusercontent.com/5323259/97776833-e1e0f880-1b6b-11eb-8275-39e6dd05b0d2.png)

One selects a layout to start with or selects to skip. The above selection is lost after having gone through the first step in selecting a layout. It is a feature that would be nice to include in a settings panel of the Columns block.
 
Mockup:
![Column-Variations-Panel](https://user-images.githubusercontent.com/5323259/97776813-a5ad9800-1b6b-11eb-87c9-ff37c7c48eb2.jpg)

This gives the option for a user to later on define the structure, and not manually have to create it.

--

The above seems like a good process of multitude of blocks that have some kind of variation.

1. Select initial structural layout (variation).
2. See the variations as a panel in the sidebar settings. 


Some blocks will also have a Style panel. This will likely look somewhat similar to a Variations panel, but they serve two different purpose. One for handling structure and the other for defining the design of that structure. One can add a few words to each panel perhaps something like: "Define/select the structure" and the other "Define/select the design" 

Styles panel is something we have become used to seeing in the sidebar settings to help us setup the design of a block. Having a predefined design to which we can select helps us understand the various design options that exist for the specific block. We can select a style to start with and then continue to modify to create the look that we are going for. The same familier method can also be brought over to Variations. We open the new panel, and we see something that reminds us of the Styles panel. But in this panel they are able to visually setup the structure. Quickly understanding what it does. It is located in the same sidebar as Styles but it serves a different purpose. As a user I would be able to quickly setup the structure and then select the style I want to use. The Styles Panel and Variations Panel are so similar to each other but have vital differences in how these affect the layout and content of a block. These should be treated similar and also be close by each other in the settings panel.

**Comment by ntsekouras** [comment]
Hey @paaljoachim - thanks for exploring this!

>It is important to have options that affects the layout (variations, styles and other options) in an obvious easy to locate place like the sidebar.

I agree with that.

>It means they can skip a variation in the initial screen and instead later on make a selection in the Variations panel.

I don't believe this issue has to be coupled with another component like `__experimentalBlockVariationPicker` that is used in Query, Columns etc..  For example the `skip` option doesn't make sense to be in every block that uses a Placeholder.

I believe this issue has just to provide a way to transform a block with one of its block variations, only if a block variation has opted in for this behavior and can also support the wanted transformation.

>The selected variation is seen in the layout and there is also a border around the selected variation in the sidebar panel. Making it obvious which variation is selected. Making it easy for the user to click through the variations and see the change in the layout.

This will not be possible as there is no way and doesn't make sense to implement to know the variation that a block had been created from (I thought so as well before and I had made a prototype recently https://github.com/WordPress/gutenberg/pull/26105, but closed it).

A simple example would be the `Query` block. We can start with `title+date` and then add `image`. So the initial variation used is invalid.

**Comment by kellychoffman** [comment]
> https://user-images.githubusercontent.com/191598/92940791-8a50c700-f41d-11ea-8f24-4b56c8e72735.png

This direction feels most promising. I think it needs to be more explicit as to what it is.

**Comment by paaljoachim** [comment]
We are now looking at two options.
A drop down or a panel.

Query block - drop down with only text.
![Query-block-Variations-2-open-drop-down](https://user-images.githubusercontent.com/5323259/98010288-cb9a9d00-1df6-11eb-97f1-9672ed61b159.jpg)

Query block - drop down with thumbnails and text.
![Query-block-Variations-3-drop-down-with-thumbnails](https://user-images.githubusercontent.com/5323259/98011511-3e584800-1df8-11eb-925d-2374f285ee2e.jpg)

----------------

Backing up for a moment.

The initial experience users will have with the Columns block looks like this:
![Screen Shot 2020-11-03 at 17 40 39](https://user-images.githubusercontent.com/5323259/98014318-bffda500-1dfb-11eb-8f39-bd1c0344c818.png)


Columns - drop down with only text.
![Column-block-sidebar-variations-drop-down](https://user-images.githubusercontent.com/5323259/98014047-6dbc8400-1dfb-11eb-9211-2eea8a91fcd8.jpg)

Columns - drop down with thumbnails and text
![Column-block-sidebar-variations-drop-down-2](https://user-images.githubusercontent.com/5323259/98015803-7b730900-1dfd-11eb-88c8-4536021f76b7.jpg)

We can see how important it is to keep the visual consistency between the placeholder/startup screen and a drop down or panel to where they can later on through thumbnails test out different structural layouts.

**Comment by ntsekouras** [comment]
> Columns - drop down with text only.

We could show the description of these variations and not just the titles.

I understand that some variations would be better understood with `icons` and we could make 2 designs conditionally by using a prop or something, but would this be consistent? 

Also since this list will contain variations that have opted-in for this behavior, we could also apply some 'rules' that will make it easily understood. For example `description` field is optional now, but we could encourage to use that field for these kind of variations (it would still be a fallback of course).
